### PR TITLE
feat(rln-relay): abstract group management into its own api

### PR DESCRIPTION
**Pull request by rymnc**
Proposes a GroupManager, which will be a field on WakuRlnRelay, having a standard interface across all implementations. This aims to solve the issues related to readability of the code when it operates in different modes.

- feat(rln-relay): group manager api

TODO (not exhaustive):
- [x] Implement event listener for onchain group manager
- [x] Move rln.nim into an ffi directory, with its wrapper procs (insertMember etc)
- [x] Move conversion utils into its own file 
- [x] More tests for static group manager
- [x] clean up utils.nim

In follow up PR:
- refactor mount to include group managers
- move constants to appropriate locations

### CODE REVIEW

**Comment by rymnc** [comment]
Note: imo we should create a separate PR to integrate these changes into the rln-relay itself. At the current moment, I have not removed any code from the utils.nim file, and left it as is for ease of review. 
It's already 1500 loc :0

**Review by LNSD — COMMENTED** [review]
Please, check my comments 🙂

File: waku/v2/protocol/waku_rln_relay.nim [hunk 1]
```diff
@@ -1,15 +1,17 @@
...
 
 export
-  rln,
+  ffi,
   constants,
   protocol_types,
   protocol_metrics,
+  conversion_utils,
   utils,
   contract
```
[the remark this answers is not in the document]
Comment (rymnc) [reply waku/v2/protocol/waku_rln_relay.nim:17]: Yes, we do test even the ffi to ensure that breaking upstream changes are detected whenever we update submodules

File: waku/v2/protocol/waku_rln_relay/conversion_utils.nim [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+  ## returns length prefixed version of the input
+  ## with the following format [len<8>|input<var>]
+  ## len: 8-byte value that represents the number of bytes in the `input`
+  ## len is serialized in little-endian
+  ## input: the supplied `input`
+  let
+    # the length should be serialized in little-endian
+    len = toBytes(uint64(input.len), Endianness.littleEndian)
+    output = concat(@len, @input)
+  return output
```
[the remark this answers is not in the document]
Comment (rymnc) [reply waku/v2/protocol/waku_rln_relay/conversion_utils.nim:54]: Will address when fixing https://github.com/waku-org/nwaku/issues/1487 :)

File: waku/v2/protocol/waku_rln_relay/conversion_utils.nim [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+  return proofBytes
+
+# Serializes a sequence of MerkleNodes
+proc serialize*(roots: seq[MerkleNode]): seq[byte] =
+  var rootsBytes: seq[byte] = @[]
+  for root in roots:
+    rootsBytes = concat(rootsBytes, @root)
+  return rootsBytes
+
+proc serializeIdCommitments*(idComms: seq[IDCommitment]): seq[byte] =
```
[the remark this answers is not in the document]
Comment (rymnc) [reply waku/v2/protocol/waku_rln_relay/conversion_utils.nim:91]: Will address when fixing https://github.com/waku-org/nwaku/issues/1487 :)

File: waku/v2/protocol/waku_rln_relay/conversion_utils.nim [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    try:
+      let
+        idSecretHash = hexToUint[IdentitySecretHash.len*8](groupKeys[i][0]).toBytesLE()
+        idCommitment = hexToUint[IDCommitment.len*8](groupKeys[i][1]).toBytesLE()
+      groupIdCredentials.add(IdentityCredential(idSecretHash: idSecretHash,
+          idCommitment: idCommitment))
+    except ValueError as err:
+      warn "could not convert the group key to bytes", err = err.msg
+      return err("could not convert the group key to bytes: " & err.msg)
+  return ok(groupIdCredentials)
```
[the remark this answers is not in the document]
Comment (rymnc) [reply waku/v2/protocol/waku_rln_relay/conversion_utils.nim:148]: Will address when fixing https://github.com/waku-org/nwaku/issues/1487 :)

File: waku/v2/protocol/waku_rln_relay/group_manager/group_manager_base.nim [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  idCommitment*: IDCommitment
+  index*: MembershipIndex
+
+type OnRegisterCallback* = proc (registrations: seq[Membership]): Future[void] {.gcsafe.}
+type OnWithdrawCallback* = proc (withdrawals: seq[Membership]): Future[void] {.gcsafe.}
+
+type GroupManagerResult*[T] = Result[T, string]
+
+type
+  GroupManager*[Config] = ref object of RootObj
```
[the remark this answers is not in the document]
Comment (rymnc) [reply waku/v2/protocol/waku_rln_relay/group_manager/group_manager_base.nim:29]: oops! looks like this change was pushed by mistake, Addressed in [4135b87](https://github.com/waku-org/nwaku/pull/1465/commits/4135b8785953b79132e144e637216cfc1c33fe3c), thanks!

File: waku/v2/protocol/waku_rln_relay.nim [hunk 6]
```diff
@@ -1,15 +1,17 @@
...
   ./waku_rln_relay/constants,
   ./waku_rln_relay/protocol_types,
   ./waku_rln_relay/protocol_metrics,
+  ./waku_rln_relay/conversion_utils,
   ./waku_rln_relay/utils,
   ./waku_rln_relay/contract
 
 export
-  rln,
+  ffi,
```
[the remark this answers is not in the document]
Comment (rymnc) [reply waku/v2/protocol/waku_rln_relay.nim:11]: Would you prefer `librln` to be more explicit?

File: waku/v2/protocol/waku_rln_relay.nim (hunk 1 above)
[the remark this answers is not in the document]
Comment (LNSD) [reply waku/v2/protocol/waku_rln_relay.nim:17]: I see. But... One thing is the module APIs, and another thing is testing.

A clean API without exposing internal symbols (types, methods, etc.) helps prevent the library's misusage that could break the program. You'll still be able to test the different submodules by importing them explicitly in the test file/module at the moment of testing.

Said that, do you still think that we should expose everything in this module?

**Comment by LNSD** [comment]
One question/suggestion regarding the module's structure: 

> Would it be possible to move the `rln` (a.k.a. `ffi` in this PR) module to the `waku/common` folder? I see the RLN wrapper as a module that should be standalone, and the `waku_rln_relay` module should depend on this common module.
> 
> When doing the exercise to move the module to the common's folder, you have crossed dependencies (`rln` depending on `waku_rln_relay` and vice-versa), then there is something in the code layout that should be reviewed. 
> 
> Disclaimer: I have not tried it. But from my experience, the likeliness is high when refactoring code and making it more modular.

File: waku/v2/protocol/waku_rln_relay.nim (hunk 6 above)
[the remark this answers is not in the document]
Comment (rymnc) [reply waku/v2/protocol/waku_rln_relay.nim]: Addressed in [0f6a3a9](https://github.com/waku-org/nwaku/pull/1465/commits/0f6a3a9bbdeb2062f4ed7f00e16eb89465337968) and [42231ae](https://github.com/waku-org/nwaku/pull/1465/commits/42231ae8c3288ab5f1b497d6272abc636233e3d7) (some gitignore shenanigans)

**Comment by rymnc** [comment]
@LNSD can I merge this now?

**Review by LNSD — APPROVED** [review]
LGTM ✅

**Pull request merged**